libraries/GCS_MAVLink/GCS_Common.cpp

```diff
@@ -815,67 +825,457 @@ bool GCS_MAVLINK::handle_mission_item(mavlink_message_t *msg, AP_Mission &missio
+        if (interval == 0) {
+            interval = 100;
+        } else if (interval > 1000) {
+            interval = 1000;
+        }
+    }
 
-    // if there are no deferred messages, attempt to send straight away:
-    if (num_deferred_messages == 0) {
-        if (try_send_message(id)) {
-            // yay, we sent it!
-            return;
+    // first find a slot for the message.  It may already have one -
+    // which we will obviously reuse, but it might not, in which case
+    // we allocate one for the message.
+    bool reusing_slot = false;
+    uint8_t free_slot = no_deferred_message_to_send;
+    uint8_t slot = no_deferred_message_to_send;
+    for (uint8_t i=0; i<ARRAY_SIZE(deferred_messages); i++) {
+        deferred_message_t &deferred_message = deferred_messages[i];
+        if (deferred_message.id == id) {
+            slot = i;
+            reusing_slot = true;
+            break;
+        }
+        if (deferred_message.id == MSG_LAST &&
+            free_slot == no_deferred_message_to_send) {
+            free_slot = i;
         }
     }
 
-    // we failed to send the message this time around, so try to defer:
-    if (num_deferred_messages == ARRAY_SIZE(deferred_messages)) {
-        // the defer buffer is full, discard this attempt to send.
-        // Note that the message *may* already be in the defer buffer
-        return;
+    if (slot == no_deferred_message_to_send) {
+        // message not found in our queue
+        if (interval == 0) {
+            // if it is not in our queue then I guess we succeeded?
+            return true;
+        }
+        // use a found free slot if possible:
+        slot = free_slot;
+    }
+    if (slot == no_deferred_message_to_send) {
+        // no free slots....
+        return false;
+    }
+
+    // don't send out a message more than once per loop.  At these
+    // sorts of rates you're probably going to start slipping
+    // messages...
+    if (interval != 0 &&
+        interval*1000 < AP::scheduler().get_loop_period_us()) {
+        interval = AP::scheduler().get_loop_period_us()/1000;
+    }
+
+    // ::fprintf(stderr, "Setting interval for %u to %ums\n", id, interval);
+
+    deferred_message_t &deferred_message = deferred_messages[slot];
+    deferred_message.id = id;
+    deferred_message.interval_ms = interval;
+    if (interval == 0) {
+        // stop sending this message and free the slot
+        deferred_message.send_not_before_ms = 0;
+        deferred_message.id = MSG_LAST;
+        next_deferred_message_to_send_state = INVALID;
+        return true;
+    }
+
+    // we avoid rescheduling for no good reason.  Otherwise any time
+    // the GCS does something like a set-stream-rate all messages come
+    // out straight away.
+    const uint32_t now = AP_HAL::millis();
+    if (!reusing_slot ||
+        deferred_message.send_not_before_ms == 0 ||
+        deferred_message.send_not_before_ms > now + interval) {
```

Comment:
this looks like it won't handle ms wrap, needs to be a subtraction for time difference